californium-core/src/main/java/org/eclipse/californium/core/network/serialization/DataSerializer.java

```diff
@@ -44,28 +45,45 @@
 	/**
 	 * Serializes a message to the wire format.
 	 * <p>
-	 * This does <em>not</em> cache the byte array in the message's <em>bytes</em> property.
+	 * This does <em>not</em> cache the byte array in the message's
+	 * <em>bytes</em> property.
 	 * 
 	 * @param message The message to serialize.
 	 * @return The encoded message.
 	 * @throws NullPointerException if message is {@code null}
+	 * @throws IllegalArgumentException if a NON empty-message is provided, or a
+	 *             empty-message uses a none-empty-token.
 	 */
 	public final byte[] getByteArray(final Message message) {
 		if (message == null) {
 			throw new NullPointerException("message must not be null!");
 		}
-		DatagramWriter optionsAndPayloadWriter = new DatagramWriter();
-		serializeOptionsAndPayload(optionsAndPayloadWriter, message.getOptions(), message.getPayload());
-		optionsAndPayloadWriter.writeCurrentByte();
-
-		MessageHeader header = new MessageHeader(CoAP.VERSION, message.getType(), message.getToken(),
-				message.getRawCode(), message.getMID(), optionsAndPayloadWriter.size());
-
 		DatagramWriter messageWriter = new DatagramWriter();
-		serializeHeader(messageWriter, header);
-		messageWriter.writeCurrentByte();
-		messageWriter.write(optionsAndPayloadWriter);
+		if (message.getRawCode() == 0) {
+			// simple serialization for empty message.
+			// https://tools.ietf.org/html/rfc7252#section-4.1
+			if (message.getType() == Type.NON) {
+				throw new IllegalArgumentException("NON message must not use code 0 (empty message)!");
+			}
+			if (!message.getToken().isEmpty()) {
+				throw new IllegalArgumentException("Empty messages must not use a token!");
+			}
+			MessageHeader header = new MessageHeader(CoAP.VERSION, message.getType(), message.getToken(), 0,
```

Comment:
Hi @boaks , LOW: it would be nice to separate serializeHeader to a new method :) rather than duplicating it in both cases

Reply:
> it would be nice to separate serializeHeader to a new method

Sure, I can reduce the duplicate of the call to `serializeHeader/writeCurrentByte`, but that would require to do more `if`s, at least for the call to `messageWriter.write(optionsAndPayloadWriter)`.
FMPOV, accepting the duplicate is the simpler approach.